elastica/modules/contact.py

```diff
@@ -0,0 +1,171 @@
+        self._contacts.append(_contact)
+
+        return _contact
+
+    def _finalize_contact(self):
+
+        # dev : the first indices stores the
+        # (first_rod_idx, second_rod_idx)
+        # to apply the contacts to
+        # Technically we can use another array but it its one more book-keeping
+        # step. Being lazy, I put them both in the same array
+        self._contacts[:] = [(*contact.id(), contact()) for contact in self._contacts]
+
+        # check contact order
+        for (
+            first_sys_idx,
+            second_sys_idx,
+            contact,
+        ) in self._contacts:
+            contact._order_check(
+                self._systems[first_sys_idx],
+                self._systems[second_sys_idx],
+            )
+
+    def _call_contacts(self, time, *args, **kwargs):
+        for (
+            first_sys_idx,
+            second_sys_idx,
+            contact,
+        ) in self._contacts:
+            contact.apply_contact(
+                self._systems[first_sys_idx],
+                self._systems[second_sys_idx],
+                *args,
+                **kwargs,
+            )
+
+
+class _Contact:
+    """
+    Contact module private class
+
+    Attributes
+    ----------
+    _first_sys_idx: int
+    _second_sys_idx: int
+    _contact_cls: list
+    *args
+        Variable length argument list.
+    **kwargs
+        Arbitrary keyword arguments.
+    """
+
+    def __init__(
+        self,
+        first_sys_idx: int,
+        second_sys_idx: int,
+    ):
+        """
+
+        Parameters
+        ----------
+        first_sys_idx
+        second_sys_idx
+        """
+        self.first_sys_idx = first_sys_idx
+        self.second_sys_idx = second_sys_idx
+        self._contact_cls = None
+        self._args = ()
+        self._kwargs = {}
+
+    def using(self, contact_cls, *args, **kwargs):
+        """
+        This method is a module to set which contact class is used to apply contact
+        between user defined rod-like objects.
+
+        Parameters
+        ----------
+        contact_cls: object
+            User defined callback class.
+        *args
+            Variable length argument list
+        **kwargs
+            Arbitrary keyword arguments.
+
+        Returns
+        -------
+
+        """
+        from elastica.contact_forces import NoContact
+
+        assert issubclass(
+            contact_cls, NoContact
+        ), "{} is not a valid contact class. Did you forget to derive from NoContact?".format(
+            contact_cls
+        )
+        self._contact_cls = contact_cls
+        self._args = args
+        self._kwargs = kwargs
+        return self
+
+    def id(self):
+        return (
+            self.first_sys_idx,
+            self.second_sys_idx,
+        )
+
+    def __call__(self, *args, **kwargs):
+        if not self._contact_cls:
+            raise RuntimeError(
+                "No contacts provided to link rod-like object id {0}"
```

Comment:
'link' is not correct term, update it

Reply:
is "to establish contact between" better?

Reply:
yes better